src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs

```diff
@@ -0,0 +1,850 @@
+                {
+                    if (prop.Value.IsDense)
+                    {
+                        arrayIndicesPointers[instanceIndex] = null;
+                    }
+                    else
+                    {
+                        int* pIndicesArray = (int*)indicesGcHandle.Value.AddrOfPinnedObject();
+                        arrayIndicesPointers[instanceIndex] = &pIndicesArray[indicesStartIndex];
+                    }
+                    float* pValuesArray = (float*)valuesGcHandle.Value.AddrOfPinnedObject();
+                    arrayValuesPointers[instanceIndex] = &pValuesArray[valuesStartIndex];
+
+                    instLabels[instanceIndex] = prop.Value.Label;
+                    instSizes[instanceIndex] = prop.Value.FeatureCount;
+                    instanceIndex++;
+                }
+
+                fixed (float* pweightVector = &weightVector[0])
+                fixed (int** pIndicesPointer = &arrayIndicesPointers[0])
+                fixed (float** pValuesPointer = &arrayValuesPointers[0])
+                fixed (int* pInstSizes = &instSizes[0])
+                fixed (float* pInstLabels = &instLabels[0])
+                {
+                    LearnAll(totalNumInstances, pInstSizes, pIndicesPointer, pValuesPointer, pInstLabels, tuneLR, ref lr, l2Const, piw,
+                            pweightVector, ref bias, numFeatres, numPasses, numThreads, tuneNumLocIter, ref numLocIter, tolerance, needShuffle, shouldInitialize, (State*)stateGCHandle.AddrOfPinnedObject());
+                }
+            }
+
+            [DllImport(DllName), SuppressUnmanagedCodeSecurity]
+            private static extern void MapBackWeightVector(float* weightVector, State* state);
+
+            /// <summary>
+            /// Maps back the dense feature to the correct position
+            /// </summary>
+            /// <param name="weightVector">The weight vector</param>
+            /// <param name="stateGCHandle"></param>
+            public static void MapBackWeightVector(float[] weightVector, GCHandle stateGCHandle)
+            {
+                fixed (float* pweightVector = &weightVector[0])
+                    MapBackWeightVector(pweightVector, (State*)stateGCHandle.AddrOfPinnedObject());
+            }
+
+            [DllImport(DllName), SuppressUnmanagedCodeSecurity]
+            private static extern void DeallocateSequentially(State* state);
+
+            public static void DeallocateSequentially(GCHandle stateGCHandle)
+            {
+                DeallocateSequentially((State*)stateGCHandle.AddrOfPinnedObject());
+            }
+        }
+
+        /// <summary>
+        /// This is the state of a SymSGD learner that is shared between the managed and native code.
+        /// </summary>
+        [StructLayout(LayoutKind.Explicit)]
+        public unsafe struct State
+        {
+#pragma warning disable 649 // never assigned
+            [FieldOffset(0x00)]
```

Comment:
Why the need to explicitly list the offsets here?  This forces us to never work on non-64-bit platforms.

Instead, we should be able to use `[StructLayout(Sequential)]` and then just list the fields as necessary.

For example:  [https://github.com/dotnet/corefx/blob/ba496ae6a9ae055708b630b3f73395a76c2ce6ce/src/System.Drawing.Common/src/System/Drawing/Internal/GpPathData.cs#L9-L15](https://github.com/dotnet/corefx/blob/ba496ae6a9ae055708b630b3f73395a76c2ce6ce/src/System.Drawing.Common/src/System/Drawing/Internal/GpPathData.cs#L9-L15)

Reply:
Given this is a port  it is probably out of scope to make this kind of change without also involving the authors of sysmsgd, instead I will open an issue and assign it to them. Does that sound good?

Reply:
That works for me.